### DISCUSSION TITLE
Dumb question about not linked commits

### DISCUSSION THREAD

**Discussion by Doomkyn** [community/community#23517] [New to GitHub]
<p>Hello everybody.<br>
I have a really dumb question.<br>
Why are unlinked commits allowed?<br>
I’m probably using github wrong, but my idea was to keep track of changes in my projects and know who make changes, but it turns out that simply changing the git config, let users push code without having them linked to the commit.<br>
This is also true in organizations. As an admin of my organization, I’d like to know who pushes code.</p>
<p>Is there a way to forbid unlinked commits, or show the account who effectively made changes?</p>

**Comment by mpboom** [comment, marked as answer]
<p>Hey <a class="mention" href="/doomkyn">@Doomkyn</a> and welcome! That is definitely not a dumb question. Unfortunately, it works this way due to how Git works. However, there are steps you can take to mitigate your concerns.</p>
<p>Every (local) installation of Git has a committer mail address set. Using that mail address, commits are made. Hence, every commit has an associated email address.</p>
<p>Then, when GitHub receives commits, it tries to tie the commit to a user by checking the commit address and seeing if there is any GitHub user with that email on their account. Usually, you’ll locally commit with <a href="mailto:foo@example.org">foo@example.org</a> and also have <a href="mailto:foo@example.org">foo@example.org</a> on your GitHub account, so GitHub can easily link the commit. <strong>Note:</strong> you can have <a href="https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-user-account/managing-email-preferences/adding-an-email-address-to-your-github-account" rel="noopener nofollow ugc">more than one email on your account</a> at the same time! By adding any used email address afterwards, you can retroactively “link” the commits.</p>
<p>Furthermore, the problem of who can push code is handled on a different level and has nothing to do with the committer mail address. You grant GitHub accounts access to push/pull on a specific repo/branch, and anyone who possesses valid credentials to that account can push commits that have any associated email address.</p>
<p>This poses a problem, however. I can locally use your <a href="mailto:foo@example.org">foo@example.org</a> address to commit, and then push it to GitHub (at least, to any repo my GitHub account was granted access to). Then, GitHub will link the commit to you while you did not commit yourself. To combat this issue, <a href="https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits" rel="noopener nofollow ugc">commit signing</a> exists. You upload your GPG key to your GitHub account and then sign your commits locally using that same key. As an administrator, you can <a href="https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#statuses-with-vigilant-mode-enabled" rel="noopener nofollow ugc">enforce commit signing</a>.</p>
<p>Hope this helps. If you got more questions, let me know.</p>

**Comment by Doomkyn** [comment]
<p>Hey <a class="mention" href="/mpboom">@mpboom</a>!<br>
Thank you very much for the detailed explaination.<br>
Everything makes sense now!</p>
<p>Have a good evening :slight_smile:</p>
